File: LeetCode/Study Plan/Practice Problems/Medium/Rectangle Area/Cpp/rectangle_area.cc
#include <iostream>

int area(int x1, int y1, int x2, int y2) { return (x2-x1)*(y2-y1); }

int rectangle_area(int ax1, int ay1, int ax2, int ay2, int bx1, int by1, int bx2, int by2) {
    int x1 = std::max(ax1, bx1);
    int y1 = std::max(ay1, by1);
    int x2 = std::min(ax2, bx2);
    int y2 = std::min(ay2, by2);
    return (x2 > x1 && y2 > y1)?area(ax1, ay1, ax2, ay2) + area(bx1, by1, bx2, by2) - area(x1, y1, x2, y2):area(ax1, ay1, ax2, ay2) + area(bx1, by1, bx2, by2);
}

int main() {
    std::cout<<rectangle_area(-2, -2, 2, 2, -2, -2, 2, 2)<<"\n";
}

File: LeetCode/Study Plan/Practice Problems/Medium/Rectangle Area/Go/rectangle_area.go
package main

import "fmt"

func rectangle_area(ax1, ay1, ax2, ay2, bx1, by1, bx2, by2 int) int {
	x1 := max(ax1, bx1)
	y1 := max(ay1, by1)
	x2 := min(ax2, bx2)
	y2 := min(ay2, by2)
	area_a1 := area(ax1, ay1, ax2, ay2)
	area_a2 := area(bx1, by1, bx2, by2)
	area_a3 := 0
	if x2 > x1 && y2 > y1 {
		area_a3 = area(x1, y1, x2, y2)
	}
	return area_a1 + area_a2 - area_a3
}

func area(x1, y1, x2, y2 int) int {
	return (x2 - x1) * (y2 - y1)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func main() {
	fmt.Println(rectangle_area(-3, 0, 3, 4, 0, -1, 9, 2))
}
